Upgraded dependencies, `composer:^2` and `php:^8`, mutation tests CI improvements

Ocramius opened the pull request:
I was mostly annoyed by the `composer install` time in CI environments, so I decided to give this a shot and update to `php:^8` and `composer/composer:^2`.

TODO:

 * [x] add CI task that performs a `docker build`, ensuring that our build process still runs

Blockers:

 * [x] https://github.com/jwage/changelog-generator/pull/64
 * [x] https://github.com/jwage/changelog-generator/pull/65
 * [x] https://github.com/phly/keep-a-changelog/issues/96
 * [x] https://github.com/phly/keep-a-changelog/issues/97

Ocramius committed:
Upgraded dependencies, running on `composer:^2` and `php:^8`

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Bumping mutation score thresholds for infection, since the codebase improved a bit

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Removed `php8.0-json`, since `ext-json` is now built into PHP 8+

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Added a daily CI pipeline that ensures that our docker image can be built even when base images change

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,21 @@
+on:
+  pull_request:
+  schedule:
+    - cron: '0 2 * * *' # run daily
```
kinda required, since ubuntu base dependencies keep shifting.

I'd be happier if I did this properly with nixos, but I need to put in the effort to actually get there myself :-)

geerteltink commented:
LGTM when the dependencies are updated. This should improve the build time a lot.

@Ocramius have you thought about building the docker container itself and push that to a registry? That way the container doesn't need to be build repeatedly and would improve speed even more. I'm not sure how that would work with actions though or if we can use GitHub registry for it. Last time I checked you could not access the GitHub registry without tokens.

geerteltink commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,21 @@
+  docker-build:
+    name: "Simulate Docker Build"
+
+    runs-on: ${{ matrix.operating-system }}
```
There is no matrix configured? Why not just `"ubuntu-latest"`?

Ocramius replied:
Good point 👍

Ocramius replied:
44344879748d773c5945a4508712c806417d0bcb

Ocramius commented:
Yeah, unsure about docker builds: would be totally for it if it were globally usable 🤔

Ocramius committed:
Upgraded dependencies, removed `composer/package-versions-deprecated`, since we're running a contemporary stack

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Run CI steps on PHP 8.0

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Removing usage of `matrix` in `docker-build.yml`, since we always `docker build` on `ubuntu-latest`

Ref: https://github.com/laminas/automatic-releases/pull/97#discussion_r555374443
Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
As per `ocramius/package-versions:^2`, `Versions::rootPackageName()` is to be used

`Versions::ROOT_PACKAGE_NAME` has been removed, see: https://github.com/Ocramius/PackageVersions/releases/tag/2.0.0

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Upgraded tooling to squish mutation test mutations where mutations don't respect static analysis

This also shows mutations for only the diffs in the code, rather than running the whole
mutation test suite (expensive).

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Ensure that the base PR branch is checked out and available for `git` commands to reference it

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Instead of messing with environment variables, use `github.base_ref`

As per https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions,
it is possible to use the `github.<...>` context to access some information about the current action
run.

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Use a simpler approach to fetching `base_ref` in mutation tests PRs

Ref: https://github.com/jasonkarns/checkout/commit/11d9aa95eb60f32a2ec2e21ded0a02a810204f5c
Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Print github-friendly mutation test report to integrate with github actions/PR UI

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius commented:
Note: the added mumbojumbo around `mutation-tests.yml` is mostly to produce decent reports (and fast ones) for users that send patches to improve small sections of the code.

That ensures that we don't need to run the full mutation test suite against everything every time, if the diff is limited to some sources.

The output looks like this: 

![Selection_567](https://user-images.githubusercontent.com/154256/105283044-f2739e80-5baf-11eb-8e76-24fe5a37bee2.png)

Ocramius commented:
Just to make sure everything went smoothly, I ran this on a pre-existing tag of `ocramius/LazyMap`, and it works fine: https://github.com/Ocramius/LazyMap/runs/1739730760?check_suite_focus=true#step:4:12 (failure expected, since tag already exists)
